packages/devtools_app/lib/src/debugger/flutter/debugger_controller.dart

```diff
@@ -462,12 +466,97 @@ class DebuggerController extends DisposableController
     _selectedStackFrame.value = frame;
 
+    if (frame?.frame != null) {
+      _variables.value = _variablesForFrame(frame.frame);
+    } else {
+      _variables.value = [];
+    }
+
     if (frame?.scriptRef != null) {
       showScriptLocation(
           ScriptLocation(frame.scriptRef, location: frame.position));
     }
   }
 
+  List<Variable> _variablesForFrame(Frame frame) {
+    final variables = frame.vars.map((v) => Variable.create(v)).toList();
+    variables.forEach(buildVariablesTree);
+    return variables;
+  }
+
+  Future<void> buildVariablesTree(Variable variable) async {
+    if (!variable.isExpandable || variable.treeInitialized) return;
+
+    final InstanceRef instanceRef = variable.boundVar.value;
+    try {
+      final dynamic result = await getObject(instanceRef);
+      if (result is Instance) {
+        if (result.associations != null) {
+          variable.addAllChildren(_associationsAsVariables(result));
+        } else if (result.elements != null) {
+          variable.addAllChildren(_elementsAsVariables(result));
+        } else if (result.fields != null) {
+          variable.addAllChildren(_fieldsAsVariables(result));
+        }
+      }
+    } on SentinelException catch (_) {
+      // Fail gracefully if calling `getObject` throws a SentinelException.
+    }
+    variable.treeInitialized = true;
+  }
+
+  List<Variable> _associationsAsVariables(Instance instance) {
+    return instance.associations
+        .map((MapAssociation assoc) {
+          // For string keys, quote the key value.
+          String keyString = assoc.key.valueAsString;
+          if (assoc.key is InstanceRef &&
+              assoc.key.kind == InstanceKind.kString) {
+            keyString = "'$keyString'";
+          }
+          return BoundVariable(
+            name: '[$keyString]',
+            value: assoc.value,
+            scopeStartTokenPos: null,
+            scopeEndTokenPos: null,
+            declarationTokenPos: null,
+          );
+        })
+        .map((bv) => Variable.create(bv))
+        .toList();
+  }
+
+  List<Variable> _elementsAsVariables(Instance instance) {
+    final List<BoundVariable> boundVars = [];
+    for (int i = 0; i < instance.elements.length; i++) {
+      final value = instance.elements[i];
+      boundVars.add(BoundVariable(
+        name: '[$i]',
+        value: value,
+        scopeStartTokenPos: null,
+        scopeEndTokenPos: null,
+        declarationTokenPos: null,
+      ));
+    }
+    return boundVars.map((bv) => Variable.create(bv)).toList();
+  }
+
+  List<Variable> _fieldsAsVariables(Instance instance) {
+    return instance.fields
+        .map((BoundField field) {
+          return BoundVariable(
+            name: field.decl.name,
+            value: field.value,
+            scopeStartTokenPos: null,
+            scopeEndTokenPos: null,
+            declarationTokenPos: null,
+          );
+        })
+        .toList()
```

Comment:
I don't think you need this 1st toList().

Reply:
removed